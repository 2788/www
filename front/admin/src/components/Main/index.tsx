import React from 'react'

import { Column } from 'libs/layout-element'
import Container from 'components/common/Container'
import * as style from './style.m.less'

export default function Main({ children }: React.PropsWithChildren<{}>) {
  return (
    <Container className={style.container}>
      <div className={style.inner}>
        <Column overflow="auto">{children}</Column>
      </div>
    </Container>
  )
}
