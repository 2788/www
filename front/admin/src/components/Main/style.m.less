.container {
  margin: 24px;
  padding: 24px;
  width: auto;
  height: 100%;
  overflow: auto;
  background-color: #ffffff;

  .inner {
    width: 100%;
    height: 100%;
  }
}