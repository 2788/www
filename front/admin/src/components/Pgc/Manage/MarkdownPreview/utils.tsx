/**
 * @file markdown preview
 * @author lizhifeng <lizhifeng@qiniu.com>
 * @description modify from `components/Www/Product/Prices/MdPreview/utils.tsx` & to be merged
 */

import React, { ReactHTML, Fragment, ReactNode } from 'react'
import unified from 'unified'
import remarkParse from 'remark-parse'
import remarkRehype from 'remark-rehype'

import { wwwDetailUrlPrefix } from 'constants/pgc'

import style from './style.m.less'

type RootNode = {
  type: 'root'
  children: Array<TextNode | ElementNode>
}

type TextNode = {
  type: 'text'
  value: string
}

type ElementNode = {
  type: 'element'
  children: Array<TextNode | ElementNode>
  properties: object
  tagName: keyof ReactHTML
}

export type RenderOptions = {
  renderExternalContent(url: string, key: number): ReactNode
}

export function mdTextToHTMLAst(text: string): Promise<RootNode> {
  return new Promise(resolve => {
    // FIXME: 这边 this 的实际类型为 Processor<void, Input, Input, Output>，
    // 但是改为实际类型比较麻烦，所以先 as any 了
    function getHTMLAstPlugin(this: any) {
      Object.assign(this, { Compiler: compiler })
      function compiler(root: RootNode) {
        resolve(root)
      }
    }

    unified()
      .use(remarkParse) // md text -> md ast
      .use(remarkRehype) // md ast -> html ast
      .use(getHTMLAstPlugin)
      .process(text)
  })
}

export function renderHTMLAst(htmlAst: RootNode, options: RenderOptions) {
  return React.createElement(
    Fragment,
    null,
    childrenToReactNode(htmlAst.children, options)
  )
}

function childrenToReactNode(
  children: Array<TextNode | ElementNode>,
  options: RenderOptions,
  parent?: ElementNode | RootNode
) {
  children = [...children]
  const res: ReactNode[] = []
  let key = 0
  for (let i = 0; i < children.length; i++) {
    const externalContentInfo = getExternalContentInfo(children, i)
    if (externalContentInfo != null) {
      res.push(options.renderExternalContent(externalContentInfo.url, key++))
      children.splice(i, externalContentInfo.end - i)
      continue
    }

    const current = children[i]

    if (current.type === 'text') {
      const text = renderTextNode(current, parent)
      if (text !== null) {
        res.push(text)
      }
      continue
    }

    if (current.type === 'element') {
      res.push(renderElementNode(current, key++, options))
      continue
    }
  }
  return res
}

interface ExternalContentInfo {
  url: string
  end: number
}

// 识别类似这种三段式结构： <hr /><p>${wwwUrlPrefix}${id}</p><hr />
function getExternalContentInfo(
  children: Array<TextNode | ElementNode>,
  index: number
): null | ExternalContentInfo {
  // 至少 3 段结构，提前熔断
  if (index + 3 > children.length) {
    return null
  }

  let current = children[index]

  // 匹配第一段
  if (!isHr(current)) {
    return null
  }

  let url: string | null = null

  for (let i = index + 1; i < children.length; i++) {
    current = children[i]

    if (isBlank(current)) {
      continue
    }

    // 匹配第二段
    if (url == null) {
      let text = ''

      if (current.type === 'text') {
        text = current.value
      } else if (current.type === 'element' && isHtmlTag(current, 'p')) {
        const childrenOfP = current.children.filter(child => !isBlank(child))
        if (childrenOfP.length === 1 && childrenOfP[0].type === 'text') {
          text = childrenOfP[0].value
        }
      }

      text = text.trim()

      if (isExternalContentUrl(text)) {
        url = text
        continue
      }

      return null
    }

    // 匹配第三段
    return isHr(current)
      ? {
        url,
        end: i
      }
      : null
  }

  return null
}

function renderElementNode(element: ElementNode, key: number, options: RenderOptions): ReactNode {
  const children = element.children
  const len = children.length
  const tagName = element.tagName.toLowerCase()
  if (tagName === 'style' || tagName === 'script') {
    return null
  }
  const reactElement = React.createElement(
    tagName,
    // TODO: 先屏蔽 style、className，后续如果要支持，可参考 react-markdown 中的做法来支持
    { key, ...element.properties, style: undefined, className: undefined },
    len !== 0 ? childrenToReactNode(children, options, element) : null
  )
  if (tagName === 'table') { // 表格外面包一层 div，防止宽度超出
    return <div className={style.tableWrapper} key={key}>{reactElement}</div>
  }
  return reactElement
}

// The table-related elements that must not contain whitespace text according to React.
const tableElements = new Set(['table', 'thead', 'tbody', 'tfoot', 'tr'])

function renderTextNode(child: TextNode, parent?: ElementNode | RootNode) {
  if (
    child.value === '\n'
    || (parent && parent.type === 'element' && tableElements.has(parent.tagName))
  ) { // 去除不必要的空白文本，React does not permit whitespace text elements as children of table
    return null
  }
  return child.value
}

// 滤掉最常见的空白符 case 以保证一定的语义健壮性
function isBlank(node: TextNode | ElementNode): boolean {
  return node.type === 'text' && node.value.match(/^\s*$/) != null // 包括 \n
    || node.type === 'element' && (
      isHtmlTag(node, 'br')
      || isHtmlTag(node, 'p') && node.children.every(child => isBlank(child))
    )
}

function isHr(node: TextNode | ElementNode): boolean {
  return node.type === 'element' && isHtmlTag(node, 'hr')
}

function isHtmlTag(element: ElementNode, tag: string): boolean {
  return element.tagName.toLowerCase() === tag.toLowerCase()
}

function isExternalContentUrl(url: string): boolean {
  return url.trim().indexOf(wwwDetailUrlPrefix) === 0
}

export function getExternalContentId(url: string): string {
  return url.trim().replace(wwwDetailUrlPrefix, '').replace(/[/?#].*/, '')
}

