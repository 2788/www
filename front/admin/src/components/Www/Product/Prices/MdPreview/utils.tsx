/**
 * @file          component  utils
 * @description   提供 mdtext -> jsx 的逻辑方法，使用类似 react-markdown 的处理逻辑，只是中间加了自定义转换逻辑（如 h3 -> tabs），
 * 具体可参考 https://github.com/unifiedjs/unified 和 https://github.com/remarkjs/react-markdown
 * @author        renpanpan
 */

import React, { ReactHTML, Fragment, ReactNode } from 'react'
import unified from 'unified'
import remarkParse from 'remark-parse'
import remarkGfm from 'remark-gfm'
import remarkRehype from 'remark-rehype'
import rehypeRaw from 'rehype-raw'

type RootNode = {
  type: 'root'
  children: Array<TextNode | ElementNode>
}

type TextNode = {
  type: 'text'
  value: string
}

type ElementNode = {
  type: 'element'
  children: Array<TextNode | ElementNode>
  properties: object
  tagName: keyof ReactHTML
}

type RenderOptions = {
  renderTabPane: (tab: ReactNode, key: string, children: ReactNode) => ReactNode
  renderTabs: (key: number, children: ReactNode) => ReactNode
  tagNames?: { [k in keyof ReactHTML]?: keyof ReactHTML } // 映射标签名称到指定标签
}

export function renderMdText(text: string, options: RenderOptions): Promise<JSX.Element> {
  return new Promise(resolve => {
    function htmlNodeToReactPlugin() {
      Object.assign(this, { Compiler: compiler })
      function compiler(root: RootNode) {
        const res = renderRootNode(root, options)
        resolve(res)
      }
    }
    unified()
      .use(remarkGfm) // 支持 github 风格的 md，如链接
      .use(remarkParse) // 将 md text 转为 md ast
      .use(remarkRehype, { allowDangerousHtml: true }) // rehype 解析树
      .use(rehypeRaw) // // rehype 再次解析树，处理 md 里面的 html
      .use([htmlNodeToReactPlugin])
      .process(text)
  })
}

function renderRootNode(root: RootNode, options: RenderOptions) {
  return React.createElement(
    Fragment,
    null,
    childrenToReactNode(root.children, options)
  )
}

function childrenToReactNode(
  children: Array<TextNode | ElementNode>, options: RenderOptions, parent?: ElementNode | RootNode
) {
  const res: ReactNode[] = []
  let key = 0
  let i = 0
  let tabPanes: ReactNode[] = []
  const len = children.length
  while (i < len) {
    const current = children[i]
    if (current.type === 'text') {
      const text = renderTextNode(current, parent)
      if (text !== null) {
        res.push(text)
      }
      i++
      continue
    }
    if (current.tagName !== 'h3') {
      res.push(renderElementNode(current, key++, options))
      i++
      continue
    }
    // h3 转为 tabs
    // tabPane 中的 tab
    const tabChildren: ReactNode[] = childrenToReactNode(current.children, options, current)
    const tab = React.createElement(Fragment, null, tabChildren)
    i++
    // tabPane 中的 content，即两个 h3 之间的内容（实际上得到的 node 中两个 h3 之间的内容还是和 h3 平级的）
    const contentChildrenSource: Array<TextNode | ElementNode> = []
    let next: TextNode | ElementNode | null = i < len ? children[i] : null
    // 当没有下一个元素或者遇到 h1|h2|h3 标签，表示到达边界，tabPane 的 content 已全部获取
    while (next !== null && (next.type === 'text' || !(/h[1-3]/.test(next.tagName)))) {
      contentChildrenSource.push(next)
      i++
      next = i < len ? children[i] : null
      continue
    }
    const contentChildren: ReactNode[] = childrenToReactNode(contentChildrenSource, options)
    const content = React.createElement(Fragment, null, contentChildren)
    tabPanes.push(options.renderTabPane(tab, `${tabPanes.length}`, content))
    // 当到达边界即 h3 三级标题结束时，tabs 中包含的 TabPane 都已经完成转换
    if (next === null || next.tagName !== 'h3') {
      res.push(options.renderTabs(key++, tabPanes))
      tabPanes = []
    }
  }
  return res
}

function renderElementNode(element: ElementNode, key: number, options: RenderOptions): JSX.Element {
  const children = element.children
  const len = children.length
  return React.createElement(
    (options.tagNames && options.tagNames[element.tagName]) || element.tagName,
    { key, ...element.properties },
    len !== 0 ? childrenToReactNode(children, options, element) : null
  )
}

// The table-related elements that must not contain whitespace text according to React.
const tableElements = new Set(['table', 'thead', 'tbody', 'tfoot', 'tr'])

function renderTextNode(child: TextNode, parent?: ElementNode | RootNode) {
  if (
    child.value === '\n'
    || (parent && parent.type === 'element' && tableElements.has(parent.tagName))
  ) { // 去除不必要的空白文本，React does not permit whitespace text elements as children of table
    return null
  }
  return child.value
}
