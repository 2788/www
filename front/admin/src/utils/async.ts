/**
 * @file 异步工具
 * @author lizhifeng <lizhifeng@qiniu.com>
 */

import { useRef, useState, useCallback } from 'react'
import { makeCancelled } from 'qn-fe-core/exception'

export function usePromise<T>() {
  const resolveRef = useRef<(value: T | PromiseLike<T>) => void>(() => undefined)
  const rejectRef = useRef<(reason?: unknown) => void>(() => undefined)
  const isPendingRef = useRef(false)

  const start = useCallback(
    () => new Promise<T>((resolve, reject) => {
      if (isPendingRef.current) {
        // eslint-disable-next-line no-console
        console.warn('Promise deprecated.')
      }
      isPendingRef.current = true

      resolveRef.current = resolve
      rejectRef.current = reject
    }).finally(() => { isPendingRef.current = false }),
    []
  )

  return {
    resolve: resolveRef.current,
    reject: rejectRef.current,
    start
  }
}

export function useModalLike<T = undefined>() {
  const [visible, setVisible] = useState(false)
  const { resolve, reject, start } = usePromise<any>()
  return {
    visible,
    resolve: useCallback((...args: T extends undefined ? [] : [T]): void => {
      setVisible(false)
      resolve(args[0])
    }, [resolve]),
    reject: useCallback((reason: unknown = makeCancelled('Modal cancelled.')): void => {
      setVisible(false)
      reject(reason)
    }, [reject]),
    open: useCallback((): Promise<T extends undefined ? void : T> => {
      setVisible(true)
      return start()
    }, [start])
  }
}
