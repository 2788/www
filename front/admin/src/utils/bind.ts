import { FieldState } from 'formstate-x'
import { RangePickerProps } from 'antd/lib/date-picker/interface'
import { CheckboxValueType } from 'antd/lib/checkbox/Group'
import { Moment } from 'moment'

export type BindRangePickerResult = Pick<RangePickerProps, 'value' | 'onChange'>

export function bindRangePicker(startState: FieldState<Moment>, endState: FieldState<Moment>): BindRangePickerResult {
  return {
    // eslint-disable-next-line no-underscore-dangle
    value: [startState._value, endState._value],
    onChange: dates => {
      const moments = dates as [Moment, Moment]
      startState.onChange(moments[0])
      endState.onChange(moments[1])
    }
  }
}

export function bindCheckboxGroup(state: FieldState<CheckboxValueType[]>) {
  return {
    // eslint-disable-next-line no-underscore-dangle
    value: state._value,
    onChange: (value: CheckboxValueType[]) => state.onChange(value)
  }
}
