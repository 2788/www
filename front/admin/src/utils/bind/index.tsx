/**
 * @file utils for formstate-x v2
 * @author nighca <nighca@live.cn>
 */

export * from './input'
export * from './form-item'
