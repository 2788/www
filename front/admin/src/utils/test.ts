import * as di from 'qn-fe-core/di'
import * as base from 'admin-base/common/utils/test'
import { defaultProvides } from 'components/Provider'

export class RendererUtils extends base.RendererUtils {
  constructor(
    provides?: di.Provides,
    containerOptions?: di.IContainerOptions,
    parentContainer?: di.IContainer
  ) {
    provides = di.mergeProvides(defaultProvides, provides)
    super(provides, containerOptions, parentContainer)
  }
}

export function createContainer(
  provides?: di.Provides,
  options?: di.IContainerOptions,
  parentContainer?: di.IContainer
) {
  provides = di.mergeProvides(defaultProvides, provides)
  return base.createContainer(provides, options, parentContainer)
}
