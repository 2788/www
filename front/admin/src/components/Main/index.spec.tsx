import React from 'react'
import { RendererUtils } from 'utils/test'

import Main from '.'

test('should render well.', () => {
  const utils = new RendererUtils()
  const renderer = utils.createWithAct(<Main />)

  expect(renderer).toMatchSnapshot()
})
