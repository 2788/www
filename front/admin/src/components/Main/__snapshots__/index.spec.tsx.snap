// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render well. 1`] = `
<div
  className=""
>
  <div>
    <div
      style={
        Object {
          "display": "flex",
          "flex": 1,
          "flexDirection": "column",
          "overflow": "auto",
        }
      }
    />
  </div>
</div>
`;
